### ISSUE TITLE
在npu多机训练时 使用lora 进行预训练qwen3.5-122b-a10b和qwen3-next-80b-a3b进行保存时报错

### ISSUE THREAD

**Issue by machaoyang1999** [modelscope/ms-swift#10101]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new bug report. / 我已经搜索过现有的 issues，确认这是一个新的 bug report。

### Bug Description / Bug 描述

在npu多机训练时 使用megatron-swift的lora 进行预训练qwen3.5-122b-a10b和qwen3-next-80b-a3b进行保存时报错 由于需要进行断点续训 所以    --no_save_optim false \
    --no_save_rng false \
具体报错信息:
[TaNkOj:AAAAAAAAAAAAAAAAAAAAAAAAAAAA
[rank0]: File "/usr/local/python3.11.15/lib/python3.11/site-packages/swift/pipelines/base.py", line 52, in main
[ranko]:result = selfruno
[rank0]:AAAAA
[rank0]: File "/usr/ocal/python3.11.15/lib/python3.11/site-packages/swift/megatron/pipelines/train/sft.py', line 68, in run[rank0]: trainertrain(train_dataset, val_dataset)[rank0]: File "/usr/ocal/python3.11.15/lb/python3.11/site-packages/swift/megatron/trainers/base.py, line 792, in train[rank0]: self.run_train_step(train_data_iterator, val_data_iterator)rank0]: File "/usr/ocal/python3.11.15/lb/python3.11/site-packages/swif/megatron/trainers/base.py" line 775, in run. train step(rank0]: self.save_checkpoint0
[rank0]: File "/usr/ocal/python3.11.15/lib/python3.11/site-packages/swif/megatron/trainers/base .py", line 839, in save checkpoint
[rank0]: async_deferred = save_mcore_checkpoint(
[rank0]:AAAAAAAAAAAAAAAA
[rank0]:AAAAAAA[ranko]:[[1]l,
rankO): async_save_request = dist checkpointing.save([TankO]:AAAAAAAAAAAAAAAAAAAAAAA
rankO]): File "/usr/ocal/python3.11.15/ilb/python3.11/site- packages/swif/megatron/utils/megatron Im utis.py". line 316, in save mcore checkpoint
[rank0]: File "/Megatron-LM/megatron/core/dist checkpointing/serialization.py", line 407, in save[rank0]: sharded_state_dict, state_dict = save_preprocess(
rank0]: File "/Megatron-LM/megatron/core/dist_checkpointing/state dict utils.py", line 55, in save_preprocess
rankOl: validate shardina intearity(
[rank0]: File "/Megatron-LM/megatron/core/dist checkpointing/validation.py', line 440, in validate sharding_ integrity
[rankO]: validate_sharding_for_key(shardings)[rankO]: File "/Megatron-LM/megatron/core/dist_ checkpointing/validation.py", line 474, in validate sharding for ke)rankO]: raise CheckpointingException
tanko megaton core dist chec pointing cre Checkpointing scption inald aces paten tor Shuadelinsordkey decodarlbyans rmpoegpertsocertieatl ora Adirlut wigh aype orhitants al pel se2048), global offset=(0.0.,0), axis fragmentations=(512,1,1), replica id=(0,0,0), prepend axis num-1, alow shpe mismatch=False.flattened_range=None) tensor

### How to Reproduce / 如何复现

ms-swift=4.6.0dev
脚本:
source /usr/local/Ascend/ascend-toolkit/set_env.sh
source /root/.bashrc

set -e
set -o pipefail

export HCCL_CONNECT_TIMEOUT=7200
#export CUDA_DEVICE_MAX_CONNECTIONS=32
#export CPU_AFFINITY_CONF=1
#export TASK_QUEUE_ENABLE=2
#export STREAMS_PER_DEVICE=32
#export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True
#export NPU_ASD_ENABLE=0
#export ASCEND_ALLOC_MEM_BY_CLIQUE=1
#export USE_MCORE_GDN=1

# 初始化变量
val_data_path=""
log_path="/npu_data/ceshi.log"
training_type=""
data_path=""
model=""
output_dir_path=""
selected_checkpoint_id=0

# 参数解析
while getopts ":f:s:d:v:m:o:l:c:" opt; do
  case $opt in
  f)
    training_type=$OPTARG
    ;;
  s)
    stage=$OPTARG
    ;;
  d)
    data_path=$OPTARG
    ;;
  v)
    val_data_path=$OPTARG
    ;;
  m)
    model=$OPTARG
    ;;
  o)
    output_dir_path=$OPTARG
    ;;
  l)
    log_path=$OPTARG
    ;;
  c)
    selected_checkpoint_index=$OPTARG
    ;;
  esac
done

if [ "${current_system_environment}" = "product" ]; then
	MASTER_HOST="$VC_WORKER_HOSTS"
	MASTER_ADDR="${VC_WORKER_HOSTS%%,*}"
	MASTER_PORT="6063"
	JOB_ID="1234"
	NNODES="$MA_NUM_HOSTS"
	NODE_RANK="$VC_TASK_INDEX"
	NPUS_PER_NODE="$MA_NUM_GPUS"
	LOGNUM=$((NNODES - 1))

elif [ "${current_system_environment}" = "test" ]; then
	WORLD_SIZE=${WORLD_SIZE}
	NNODES=$((${WORLD_SIZE} / ${LOCAL_WORLD_SIZE}))
	NPUS_PER_NODE=${LOCAL_WORLD_SIZE}
	MASTER_ADDR=${MASTER_ADDR}
	MASTER_PORT=${MASTER_PORT}
	NODE_RANK=${RANK}
	LOGNUM=$((NNODES - 1))
else
	echo "Error: Unsupported current_system_environment value: ${current_system_environment}"
	echo "Supported values: product, test"
	exit 1
fi

echo "MASTER_ADDR                                        ${MASTER_ADDR}"
echo "MASTER_PORT                                        ${MASTER_PORT}"
echo "NNODES                                             ${NNODES}"
echo "NODE_RANK                                          ${NODE_RANK}"
echo "NPUS_PER_NODE                                      ${NPUS_PER_NODE}"
echo "LOGNUM                                             ${LOGNUM}"

echo "training_type: ${training_type}"
echo "epoch: ${MAAS_epoch}"
echo "data_path: ${data_path}"
echo "val_data_path: ${val_data_path}"
echo "log_path: ${log_path}"
echo "model_path: ${model}"
echo "output_dir_path: ${output_dir_path}"

echo "将文件路径分隔符由逗号改为空格"
echo "修改前data_path:${data_path}"
data_path="${data_path//,/ }"  # 直接替换所有逗号为空格
echo "修改后data_path:${data_path}"
echo "修改前val_data_path:${val_data_path}"
val_data_path="${val_data_path//,/ }"  # 直接替换所有逗号为空格
echo "修改后val_data_path:${val_data_path}"

# 计算所有文件的总行数
total_lines=0
for file in ${data_path}; do
    if [ -f "${file}" ]; then
        lines=$(wc -l < "${file}")
        total_lines=$((total_lines + lines))
    else
        echo "警告: 文件不存在 - ${file}" >&2
    fi
done
echo "total_lines:${total_lines}"
# 使用总行数计算 TRAINITERS
TRAINITERS=$(awk "BEGIN {print int((${total_lines} * ${MAAS_epoch}) / ${MAAS_global_batch_size})}")

# 构建基础命令
CMD="NNODES=${NNODES} NODE_RANK=${NODE_RANK} MASTER_ADDR=${MASTER_ADDR} MASTER_PORT=${MASTER_PORT} NPROC_PER_NODE=${NPUS_PER_NODE} megatron pt \
    --model ${model} \
    --save_safetensors true \
    --add_version false \
    --dataset ${data_path}"

# 如果 val_data_path 非空，添加 --val_dataset 参数
if [ -n "${val_data_path}" ]; then
    CMD="${CMD} \
    --val_dataset ${val_data_path}"
fi

#if [ "${training_type}" == "full" ]; then
#    echo "training_type:${training_type} 使用重计算"
#    CMD="${CMD} \
#        --recompute_granularity full \
#        --recompute_method uniform \
#        --recompute_num_layers 1"
#fi

if [ "${training_type}" == "full" ] || [ "${training_type}" == "lora" ]; then
    echo "training_type:${training_type} 使用重计算"
    CMD="${CMD} \
        --recompute_granularity full \
        --recompute_method uniform \
        --recompute_num_layers 1"
fi

is_finetune=true
# 断点续训
if [ "$selected_checkpoint_index" -ne 0 ]; then
    is_finetune=false
    megatron_model="${output_dir_path}/checkpoint-${selected_checkpoint_index}"
    # 根据training_type处理checkpoint
    if [ "${training_type}" == "lora" ]; then
        CMD="${CMD} \
        --mcore_adapter ${megatron_model}"
        echo "--mcore_adapter:${megatron_model}"
    elif [ "${training_type}" == "full" ]; then
        CMD="${CMD} \
        --mcore_model ${megatron_model}"
        echo "--mcore_model:${megatron_model}"
    else
        echo "错误: 不支持的training_type: ${training_type}"
        exit 1
    fi
fi
echo "is_finetune:${is_finetune}"

# 添加其余参数
CMD="${CMD} \
    --attention_backend flash \
    --tuner_type ${training_type} \
    --split_dataset_ratio ${MAAS_split} \
    --gradient_accumulation_fusion false \
    --train_iter ${TRAINITERS} \
    --eval-iters ${MAAS_eval_iters} \
    --micro_batch_size ${MAAS_micro_batch_size} \
    --global_batch_size ${MAAS_global_batch_size} \
    --tensor_model_parallel_size ${MAAS_tensor_model_parallel_size} \
    --pipeline_model_parallel_size ${MAAS_pipeline_model_parallel_size} \
    --sequence_parallel true \
    --moe_permute_fusion true \
	  --moe_shared_expert_overlap true \
		--moe_aux_loss_coeff 1e-3 \
		--expert_model_parallel_size ${MAAS_expert_model_parallel_size} \
		--padding_free false \
    --packing false \
    --finetune ${is_finetune} \
    --lr ${MAAS_lr} \
    --lr_warmup_iters 300 \
    --lr_warmup_fraction ${MAAS_lr_warmup_fraction} \
    --min_lr ${MAAS_min_lr} \
    --cross_entropy_loss_fusion true \
    --eval_steps ${MAAS_eval_interval} \
    --eval_iters ${MAAS_eval_iters} \
    --save_steps ${MAAS_save_interval} \
    --save_total_limit ${MAAS_save_total_limit} \
    --logging_steps ${MAAS_log_interval} \
    --max_length ${MAAS_seq_length} \
    --output_dir ${output_dir_path} \
    --no_save_optim false \
    --no_save_rng false \
    --dataset_num_proc 4 \
    --dataloader_num_workers 4"

# 执行训练命令
echo "开始训练..."
# 只有 rank0 节点记录日志，其他节点不记录
if [ "$NODE_RANK" -eq 0 ]; then
    eval ${CMD} 2>&1 | tee -a ${log_path}
else
    eval ${CMD} 2>&1  # 其他节点只输出到控制台，不写文件
fi

# 检查训练是否成功
if [ ${PIPESTATUS[0]} -ne 0 ]; then
    echo "训练失败"
    exit 1
fi

if [ "$NODE_RANK" -eq 0 ]; then
  # 训练完成后，查找并合并checkpoint
  echo "训练完成, 开始查找并合并checkpoint..."

  # 在子目录下查找checkpoint-*格式的目录，选择数字最小的
  checkpoint_dir=$(find ${output_dir_path} -maxdepth 1 -type d -name "checkpoint-*" | \
      grep -v "checkpoint-[0-9]*-merged$" | \
      sed 's/.*\/checkpoint-\([0-9]*\)/\1 &/' | \
      sort -k1 -n | \
      head -n 1 | \
      awk '{print $2}')

  if [ -z "${checkpoint_dir}" ]; then
      echo "错误: 在 ${output_dir_path} 中未找到 checkpoint-* 格式的目录"
      exit 1
  fi

  echo "找到checkpoint: ${checkpoint_dir}"

  # 根据training_type处理checkpoint
  if [ "${training_type}" == "lora" ]; then
      # Lora类型
      echo "training_type为lora，使用checkpoint-merged路径"
      merged_model_path="${checkpoint_dir}-merged"
      echo "合并后的模型路径: ${merged_model_path}"
  elif [ "${training_type}" == "full" ]; then
      # Full类型：直接使用checkpoint_dir
      echo "training_type为full，直接使用checkpoint路径..."
      merged_model_path="${checkpoint_dir}"
      echo "full模型路径: ${merged_model_path}"
  else
      echo "错误: 不支持的training_type: ${training_type}"
      exit 1
  fi

  # 将merged_model_path输出到文件
  best_model_file="${output_dir_path}/best_model_dir_path.txt"
  echo "${merged_model_path}" > ${best_model_file}
  echo "最佳模型路径已保存到: ${best_model_file}"

  # 将merged_model_path的内容复制到output_dir_path
  echo "开始移动模型到输出目录..."

  # 复制merged_model_path目录下的所有内容到output_dir_path 使用 -f 参数强制覆盖同名文件 进入源目录，使用 find 查找所有非 iter_* 的文件和目录
  cd ${merged_model_path}
  find . -maxdepth 1 ! -name "." ! -name "iter_*" -exec mv -f {} ${output_dir_path}/ \; 2>&1 | tee -a ${log_path}
  if [ ${PIPESTATUS[0]} -ne 0 ]; then
      echo "错误: 模型移动失败"
      exit 1
  fi
  echo "模型已移动到: ${output_dir_path}"

  # 更新best_model_file为output_dir_path
  echo "${output_dir_path}" > ${best_model_file}
  echo "最佳模型路径已更新为: ${output_dir_path}"

  # 后续推理可以使用 ${output_dir_path} 变量
  echo "最终模型路径: ${output_dir_path}"

  echo "开始删除 checkpoint-* 临时目录..." | tee -a ${log_path}
  # 直接删除所有匹配的目录，并检查结果
  rm -rf ${output_dir_path}/checkpoint-* 2>&1 | tee -a ${log_path}
  if [ $? -eq 0 ]; then
      echo "所有 checkpoint-* 目录已删除" | tee -a ${log_path}
  else
      echo "警告: 部分目录删除失败" | tee -a ${log_path}
  fi
  # 脚本正常结束，trap会自动调用cleanup发送completed=1
fi

**Comment by addsubmuldiv** [comment]
看起来是 NPU MoE LoRA 保存 MCore checkpoint 时误用普通 TP rank 标记专家权重副本，导致部分专家分片被判定为缺失，我在单机上修复了，但未经过多机验证，可以试一下 modelscope/mcore-bridge#196

**Comment by machaoyang1999** [comment]
好的 谢谢哥
